File: go-backend-interview-prep/08_http_and_backend/08_auth_basics_api_key/main.py
"""Auth basics (API key) -- Python equivalent of the Go example.

Simple header-based API key check. This is a toy -- production uses JWT/OAuth.
"""

import hmac
import json
import threading
import urllib.request
import urllib.error
from http.server import HTTPServer, BaseHTTPRequestHandler

API_KEY = "secret123"


def json_response(handler, status: int, data: dict) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json")
    handler.end_headers()
    handler.wfile.write(json.dumps(data).encode())


def check_api_key(handler) -> bool:
    """Check X-API-Key header. Returns True if valid."""
    key = handler.headers.get("X-API-Key", "")
    # Constant-time comparison (like Go's subtle.ConstantTimeCompare)
    return hmac.compare_digest(key, API_KEY)


class Handler(BaseHTTPRequestHandler):
    def do_GET(self):
        # Public endpoint -- no auth
        if self.path == "/health":
            json_response(self, 200, {"status": "ok"})
            return

        # Protected endpoint -- requires API key
        if self.path == "/protected":
            if not check_api_key(self):
                json_response(self, 401, {
                    "error": "unauthorized: invalid or missing API key",
                })
                return
            json_response(self, 200, {
                "message": "welcome, authenticated user!",
                "secret": "the answer is 42",
            })
            return

        json_response(self, 404, {"error": "not found"})

    def log_message(self, *args):
        pass


def main() -> None:
    server = HTTPServer(("127.0.0.1", 0), Handler)
    host, port = server.server_address
    base = f"http://{host}:{port}"
    print(f"Server on {base}")
    print("  GET /health     -> public (no auth)")
    print("  GET /protected  -> requires X-API-Key header")
    print(f"  API key: {API_KEY}")
    threading.Thread(target=server.serve_forever, daemon=True).start()

    print("\n--- Demo ---")

    # Public endpoint
    resp = urllib.request.urlopen(f"{base}/health", timeout=5)
    print(f"  /health (no key)    -> {resp.status}: {resp.read().decode()}")

    # Protected: no key -> 401
    try:
        urllib.request.urlopen(f"{base}/protected", timeout=5)
    except urllib.error.HTTPError as e:
        print(f"  /protected (no key) -> {e.code}: {e.read().decode()}")

    # Protected: wrong key -> 401
    req = urllib.request.Request(
        f"{base}/protected",
        headers={"X-API-Key": "wrong"},
    )
    try:
        urllib.request.urlopen(req, timeout=5)
    except urllib.error.HTTPError as e:
        print(f"  /protected (wrong)  -> {e.code}: {e.read().decode()}")

    # Protected: valid key -> 200
    req2 = urllib.request.Request(
        f"{base}/protected",
        headers={"X-API-Key": API_KEY},
    )
    resp2 = urllib.request.urlopen(req2, timeout=5)
    print(f"  /protected (valid)  -> {resp2.status}: {resp2.read().decode()}")

    server.shutdown()
    print("\nDone.")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/08_http_and_backend/08_auth_basics_api_key/main.go
package main

import (
	"crypto/subtle"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"net/http"
	"time"
)

// apiKey would come from env in production
const apiKey = "secret123"

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

// apiKeyAuth middleware checks X-API-Key header.
func apiKeyAuth(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		key := r.Header.Get("X-API-Key")

		// Constant-time comparison prevents timing attacks
		if subtle.ConstantTimeCompare([]byte(key), []byte(apiKey)) != 1 {
			writeJSON(w, http.StatusUnauthorized, map[string]string{
				"error": "unauthorized: invalid or missing API key",
			})
			return
		}
		next.ServeHTTP(w, r)
	})
}

func protectedHandler(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{
		"message": "welcome, authenticated user!",
		"secret":  "the answer is 42",
	})
}

func healthHandler(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
}

func main() {
	// Protected routes (behind auth)
	protectedMux := http.NewServeMux()
	protectedMux.HandleFunc("/protected", protectedHandler)

	// Public routes (no auth)
	mainMux := http.NewServeMux()
	mainMux.HandleFunc("/health", healthHandler)
	mainMux.Handle("/protected", apiKeyAuth(protectedMux))

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		log.Fatal(err)
	}
	addr := ln.Addr().String()
	fmt.Printf("Server on http://%s\n", addr)
	fmt.Println("  GET /health     -> public (no auth)")
	fmt.Println("  GET /protected  -> requires X-API-Key header")
	fmt.Printf("  API key: %s\n", apiKey)

	server := &http.Server{Handler: mainMux}

	// Self-test
	go func() {
		time.Sleep(100 * time.Millisecond)
		base := "http://" + addr
		fmt.Println("\n--- Demo ---")

		// Public endpoint
		resp, _ := http.Get(base + "/health")
		if resp != nil {
			var body map[string]string
			json.NewDecoder(resp.Body).Decode(&body)
			resp.Body.Close()
			fmt.Printf("  /health (no key)  -> %d: %v\n", resp.StatusCode, body)
		}

		// Protected: no key
		resp2, _ := http.Get(base + "/protected")
		if resp2 != nil {
			var body map[string]string
			json.NewDecoder(resp2.Body).Decode(&body)
			resp2.Body.Close()
			fmt.Printf("  /protected (no key) -> %d: %v\n", resp2.StatusCode, body)
		}

		// Protected: wrong key
		req, _ := http.NewRequest("GET", base+"/protected", nil)
		req.Header.Set("X-API-Key", "wrong")
		resp3, _ := http.DefaultClient.Do(req)
		if resp3 != nil {
			var body map[string]string
			json.NewDecoder(resp3.Body).Decode(&body)
			resp3.Body.Close()
			fmt.Printf("  /protected (wrong)  -> %d: %v\n", resp3.StatusCode, body)
		}

		// Protected: correct key
		req2, _ := http.NewRequest("GET", base+"/protected", nil)
		req2.Header.Set("X-API-Key", apiKey)
		resp4, _ := http.DefaultClient.Do(req2)
		if resp4 != nil {
			var body map[string]string
			json.NewDecoder(resp4.Body).Decode(&body)
			resp4.Body.Close()
			fmt.Printf("  /protected (valid)  -> %d: %v\n", resp4.StatusCode, body)
		}

		fmt.Println("\nDemo done.")
		server.Close()
	}()

	if err := server.Serve(ln); err != http.ErrServerClosed {
		log.Fatal(err)
	}
}
